raft: add entryID and logSlice types

pav-kv opened the pull request:
This PR introduces type-safe wrappers for log entry IDs and raft log slices. We will use them for a more readable and safe code. Some usage is introduced in this PR, see individual commits.

A lot of code in this repo manipulates `(term, index)` tuples as separate raw ints. Likewise, entry slices are handled without association with the leader term. Getting this usage wrong can be costly, since both entry IDs and log slices are centrepiece of raft safety.

Related to: #142, #144

pav-kv commented:
@ahrtr @nvanbenschoten This addresses one of the feedback points from #139 review.

PTAL. Commit-by-commit review might be the most convenient.

pav-kv commented on types.go:
```diff
@@ -0,0 +1,50 @@
+// does not necessarily match entryID.term of the entries being appended, since
+// a leader can send any entries from earlier terms in its own log.
+//
+// TODO(pav-kv): add entries, and helpers for validating the invariants of the append,
```
After we add `entries` to this, we can rename `logAppend` to a more generic `logSlice`. For **any** slice of the log, all these invariants are true, and all the fields are useful.

Specifically, the `unstable` log is a `logSlice` in which `prev` is the snapshot entry ID. The in-storage log is the physical counterpart of `logSlice`.

We can then consolidate most of the append arithmetics in `unstable` into this struct. This makes it a nice unit, that we can test properly without needing the boilerplate for setting up `unstable`/`raftLog` etc.

pav-kv replied:
Done in this PR.

pav-kv commented on types.go:
```diff
@@ -0,0 +1,50 @@
+//
+// Every entry is associated with a leadership term which issued this entry and
+// initially appended it to the log. There can only be one leader at any term,
+// and a leader never issues entries with the same index.
```
s/never issues entries/never issues two entries/

nvanbenschoten commented on types.go:
```diff
@@ -0,0 +1,50 @@
+// initially appended it to the log. There can only be one leader at any term,
+// and a leader never issues entries with the same index.
+type entryID struct {
+	term  uint64
+	index uint64
```
I think what you have here is better, so that a reader just looking at the struct definition can understand.

pav-kv commented on raft.go:
```diff
@@ -1730,16 +1730,19 @@ func stepFollower(r *raft, m pb.Message) error {
 	}
 	r.logger.Debugf("%x [logterm: %d, index: %d] rejected MsgApp [logterm: %d, index: %d] from %x",
-		r.id, r.raftLog.zeroTermOnOutOfBounds(r.raftLog.term(m.Index)), m.Index, m.LogTerm, m.Index, m.From)
+		r.id, r.raftLog.zeroTermOnOutOfBounds(r.raftLog.term(a.prev.index)), a.prev.index, a.prev.term, a.prev.index, m.From)
```
Yeah, I guess that's true. I was hoping to reduce the printing boilerplate - all these index/term pairs can be printed as a single unit. But it's a good point that at this level we should be printing the message input.

pav-kv commented on types.go:
```diff
@@ -0,0 +1,50 @@
+// initially appended it to the log. There can only be one leader at any term,
+// and a leader never issues entries with the same index.
+type entryID struct {
+	term  uint64
```
This will potentially blow up the size of this PR. Let's keep this TODO in mind.

pav-kv replied:
Added a TODO in comment.

ahrtr replied:
I tend to keep it as it's. It adds unnecessary complexity (type cast) and accordingly reduce the readability.  The naming `term` and `index` are already good enough to self-describe the meaning.

pav-kv replied:
Removed the TODO.

pav-kv commented:
@nvanbenschoten Addressed your comments, thank you. PTAL.

pav-kv commented on types.go:
```diff
@@ -0,0 +1,50 @@
+	index uint64
+}
+
+// logAppend is the metadata that describes a validated raft log append.
```
I.e. the semantics of these appends are "from leader to log".
When this is generalized to a `logSlice`, the semantics will be "a sub-slice of the leader log at `logSlice.term`".

pav-kv replied:
The leader appends happen [here](https://github.com/pav-kv/raft/blob/d14e61bd82582490766aef1d4b13986fa2573a53/raft.go#L797-L813). That `raftLog.append` method should probably be [unified with](https://github.com/pav-kv/raft/blob/main/log.go#L107-L140) `maybeAppend` into one.

For a leader appending to its own log, the safety check on `Term` would trivially be true.

pav-kv replied:
Generalized the `logAppend` struct to be `logSlice` in this PR. It will be a central part of a few more clean-ups, so it's good lay the right foundation.

pav-kv commented on types.go:
```diff
@@ -0,0 +1,52 @@
+
+// logAppend is the metadata that describes a validated raft log append.
+//
+// Every log append is done on behalf of a leader at a specific term. This term
```
`s/a leader/an elected leader/`

This is a crucial bit that distinguishes `raft.Term` from the `term` in this struct. The `term` in this struct necessarily means that there is a leader at this term.

pav-kv commented:
@nvanbenschoten @ahrtr Things start stacking up on top of this PR. Please let me know if there is major objection to this clean-up. If not, I would like to merge and iterate/rebase.

pav-kv commented on log.go:
```diff
@@ -452,12 +452,12 @@ func (l *raftLog) matchTerm(id entryID) bool {
-	if maxIndex > l.committed && term != 0 && l.zeroTermOnOutOfBounds(l.term(maxIndex)) == term {
-		l.commitTo(maxIndex)
+func (l *raftLog) maybeCommit(at entryID) bool {
+	// NB: term should never be 0 on a commit because the leader campaigned at
```
This `term != 0` check is probably unnecessary, but leaving as is in case there is some corner case that doesn't handle this properly. There should be an invariant that well protects us here.

nvanbenschoten reviewed (approved):
Looks great!

pav-kv commented:
Thanks!
@ahrtr Could you take a final pass?

pav-kv commented:
@nvanbenschoten I've changed this PR a little since the last review. Mainly, I reworked the `logSlice` struct, and improved commenting for it. Otherwise the change is equivalent.

@ahrtr @serathius Could you please take a final look at this PR, and merge if it looks good? Commit-by-commit review can be most convenient. I have follow-up work that propagates safe struct usage in other places, but I am not including them in this PR to keep it small and simple. Thank you.

ahrtr commented on types.go:
```diff
@@ -0,0 +1,108 @@
+	prev := a.prev
+	for i := range a.entries {
+		id := pbEntryID(&a.entries[i])
+		if id.term < prev.term || id.index != prev.index+1 {
+			return fmt.Errorf("entries %+v and %+v not consistent", prev, id)
+		}
```
```suggestion
		if id.term < prev.term || id.term > a.term || id.index != prev.index+1 {
			return fmt.Errorf("entries %+v and %+v not consistent, leader term: %d", prev, id, a.term)
		}
```

pav-kv replied:
We don't need to compare every entry `term` with `a.term`. The last entry has the biggest term, so it's enough to only check the last one (see a couple of lines below).

pav-kv replied:
Modified the error message to contain the leader term though, this would be useful for debugging.

ahrtr commented on types.go:
```diff
@@ -0,0 +1,108 @@
+// is sourced from a message that was received via transport, or from Storage,
+// or in a test code that manually hard-codes this struct. In these cases, the
+// invariants should be validated using the valid() method.
+type logSlice struct {
+	// term is the leader term containing the given entries in its log.
+	term uint64
+	// prev is the ID of the entry immediately preceding the entries.
+	prev entryID
+	// entries contains the consecutive entries representing this slice.
+	entries []pb.Entry
+}
```
I am thinking it might be better to replace the following fields in Message with this struct.

https://github.com/etcd-io/raft/blob/2f109970c68e505c223f4d3d5c13a8dcd8f7d528/raftpb/raft.pb.go#L402-L415

This can be addressed in separate PRs. In order to be backward compatible, we might want to do it across multiple releases. e.g. add logSlice into protobuf firstly, and then replace the fields in Message in a following release.

pav-kv replied:
Yeah, we should probably leave this for some future consideration. The scope of this work currently is internal cleanup that doesn't affect public API / wire protocol.

There are a few additional considerations:

- `logSlice` is intended as a code-level data structure. It's usually a bad practice to use protobuf-generated structs as data structures because we have limited control of its content, and also this is part of the public API.
- The `LogTerm/Index` fields are used by a bunch of other message types (not only `MsgApp`). We can't easily deprecate them, rather it would have to be a new `LogSlice` field or something. A broader refactoring could be: split the `Message` into message-type-specific sub-structs.

There is some argument for moving in a slightly opposite direction from protobufs:
- `raft` doesn't need to know every `pb.Entry`, it only internally cares about config changes. Exposing application-specific entry contents to this package poses security and privacy risks. There could be a world in which `raft` is purely a code-level algorithm/protocol, and messages are packaged at application level.
- `raft` necessitates dependency on protobuf, which is not the most efficient format: it has to be marshaled/unmarshaled. Some applications may choose other wire formats, for instance `flatbuffers`.

ahrtr commented:
A generic comment, I prefer not to add too much small commits, as most of them are mechanical changes, e.g. `use xxx in xxx`. So suggest to squash the commits, or get the `type.go and types_test.go` in a commit, and get all others in another commit.

ahrtr commented on types.go:
```diff
@@ -0,0 +1,108 @@
+// the terms of the entries in a log never regress. Property (4) means that a
+// leader log at a specific term never has entries from higher terms.
+//
+// Users of this struct can assume the invariants hold true. Exception is the
+// "gateway" code that initially constructs logSlice, such as when its content
+// is sourced from a message that was received via transport, or from Storage,
+// or in a test code that manually hard-codes this struct. In these cases, the
+// invariants should be validated using the valid() method.
```
Can you explain more about the exceptions? I think we should add a generic verification to validate the invariant properties in each test if possible.

pav-kv replied:
Agree with that. This PR adds a `valid()` check to `TestLogMaybeAppend`, which is at the moment the only user of the new struct. More tests will be converted in a follow-up PR. The `valid()` check in other tests helped to find a few places where these invariants are violated (so the tests are testing incorrect behaviour).

pav-kv replied:
The "exception" is basically any place which receives a log slice from an untrusted source (message, storage, etc). Example: see the `handleAppendEntries` call: https://github.com/etcd-io/raft/blob/2c8980be8c0d4a9aded784913a1ec5e07f3c1cce/raft.go#L1748-L1750

Once the invariants are validated, this struct can be passed around internally, and all the code can assume invariants are true. For instance, we know that the `i`-th entry in `entries` slice necessarily has `entries[i].index == prev.index + 1 + i`, so we don't have to additionally check for this.

I think the comment already explains this well, but please suggest a better phrasing if it doesn't.

pav-kv commented on types.go:
```diff
@@ -0,0 +1,108 @@
+	return a.prev
+}
+
+// valid returns nil iff the logSlice is a well-formed log slice. See logSlice
```
"iff" stands for "[if and only if](https://en.wikipedia.org/wiki/If_and_only_if)", and communicates the semantics more accurately than "if"

pav-kv commented:
@ahrtr
> A generic comment, I prefer not to add too much small commits, as most of them are mechanical changes

I prefer to isolate mechanical changes to separate commits, so that all changes have a small scope and are obviously correct. Such a PR "tells a story" rather than dumps a large change onto a reviewer. Small changes make it possible to review one thing at a time, and not have to do mental work to decompose it. I do this both for the reviewer and for myself, to self-review and convince myself too.

I usually communicate in the PR that commit-by-commit may be convenient / increase confidence in this being a mechanical change. This is not mandatory to review things this way though - if you prefer the other way around feel free to review the whole change rather than individual commits.

pav-kv commented:
@ahrtr Addressed actionable comments. Thanks for the review!

ahrtr reviewed (approved):
LGTM with a minor comment

Great work. Thanks.

pav-kv commented on raft.go:
```diff
@@ -1745,12 +1745,15 @@ func stepFollower(r *raft, m pb.Message) error {
-	if prev.index < r.raftLog.committed {
+	// TODO(pav-kv): construct logSlice up the stack next to receiving the
+	// message, and validate it before taking any action (e.g. bumping term).
+	a := logSlice{term: m.Term, prev: entryID{term: m.LogTerm, index: m.Index}, entries: m.Entries}
```
Added a `logSliceFromMsgApp` func, because it should specifically be a `MsgApp` message.

ahrtr replied:
> because it should specifically be a `MsgApp` message.

In that case, we should add a verification each time when the function is called in case it's misused.. The verification should only be executed in test. Let me add a generic verification framework/utility in a separate PR.
